### ISSUE TITLE
Issues with run_cimoperation.py and subscription manager

### ISSUE THREAD

**Issue by KSchopmeyer** [pywbem/pywbem#616]
Whoops, my bad. did not run run_cimoperations against latest code.  So here are some issues:

1.  The new code calls six.range inseveral places.  The correct code is six.moves.range.

2. The new code tests for no servers as follows (line 4381 and 4391)
    ```
            sub_mgr.remove_all_servers()
            self.assertEqual(set(sub_mgr._servers.keys()), {})
    ```
    This fails as follows:
    ```
        self.assertEqual(set(sub_mgr._servers.keys()), {})
    AssertionError: set() != {}
    ```

    **Andy:** Why use the set for this test since the assert is for an empty list? Second, the results of the set are not  an empty dictionary, but an empty list. 

    **AM:** Karl, I wanted to allow for different orderings in the result, hence the set. It is true that keys() returns a list, therefore the cast to set(). Note that {1,2,3} is also a set. However, {} is a dict, which I did not realize... So in case of the empty result, we should check for emptiness of the dict: 
    ```
    self.assertFalse(sub_mgr._servers)
    ```
    which I personally don't like because each time I need to stare at it for a while until I get what it tests for and whether it expects empty or non-empty, but it is considered more pythonic than this:
    ```
    self.assertEqual(len(sub_mgr._servers), 0)
    ```
    `</AM>`

3. The test on context manager at line 4386 fails as follows:
    ```
    Traceback (most recent call last):
      File "testsuite/run_cim_operations.py", line 4389, in test_subscription_context_manager
        server_id = sub_mgr.add_server(server)
    AttributeError: 'NoneType' object has no attribute 'add_server'
    ```
    The issue here is the `__enter__`. It must return the object that is the context.

    ```
    def __enter__(self):
        """
        Enter method when the class is used as a context manager.
        """
        return self
    ```

    **AM:** Correct. Silly me. I thought I had that right ... I'll fix that.

4. PyWBEMListenerClass.test_both_indications test function returns
    ```
    ERROR: test_both_indications (__main__.PyWBEMListenerClass)
    Create a server, listener, etc. create a filter  and subscription
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "testsuite/run_cim_operations.py", line 4305, in test_both_indications
        server_id, filter_path_owned)
      File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/pywbem-0.10.0.dev0-py3.4.egg/pywbem/_subscription_manager.py", line 954, in add_subscriptions
        server_id, filter_path, dest_path, owned)
      File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/pywbem-0.10.0.dev0-py3.4.egg/pywbem/_subscription_manager.py", line 975, in add_subscriptions
        owned)
      File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/pywbem-0.10.0.dev0-py3.4.egg/pywbem/_subscription_manager.py", line 1312, in _create_subscription
        sub_path = server.conn.CreateInstance(sub_inst)
      File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/pywbem-0.10.0.dev0-py3.4.egg/pywbem/cim_operations.py", line 3987, in CreateInstance
        **extra)
      File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/pywbem-0.10.0.dev0-py3.4.egg/pywbem/cim_operations.py", line 891, in _imethodcall
        raise CIMError(code, err[1]['DESCRIPTION'])
    pywbem.exceptions.CIMError: 11: CIM_ERR_ALREADY_EXISTS: root/PG_InterOp:CIM_IndicationSubscription.Filter="root/PG_InterOp:CIM_IndicationFilter.CreationClassName=\"CIM_IndicationFilter\",Name=\"pywbemfilter:owned:testNotOwned:owned:4da947f9-ea37-41c4-bd0c-5f7a99231e52\",SystemCreationClassName=\"CIM_ComputerSystem\",SystemName=\"\"",Handler="root/PG_InterOp:CIM_ListenerDestinationCIMXML.CreationClassName=\"CIM_ListenerDestinationCIMXML\",Name=\"pywbemdestination:owned:testNotOwned:069f24a5-e13b-4269-b11b-a0ce9b277a33\",SystemCreationClassName=\"CIM_ComputerSystem\",SystemName=\"\""
    ```

    We are creating a duplicate subscription even in the simplest test but I cannot figure out exactly where it is coming from.

    This happens even with the simplest test 
    `testsuite/run_cim_operations.py --yamlfile subscripterror.yaml http://localhost ks%ks PyWBEMListenerClass.test_create_delete_subscription`

    ```
           my_listener = self.create_listener(http_port=http_listener_port,
                                               https_port=https_listener_port)

            sub_mgr = WBEMSubscriptionManager(subscription_manager_id='fred')

            server_id = sub_mgr.add_server(server)
            listener_url = '%s://%s:%s' % ('http', 'localhost',
                                           http_listener_port)
            sub_mgr.add_listener_destinations(server_id, listener_url)

            filter_ = sub_mgr.add_filter(server_id,
                                         test_class_namespace,
                                         test_query,
                                         query_language="DMTF:CQL",
                                         filter_id="NotUsed")
            filter_path = filter_.path

            subscriptions = sub_mgr.add_subscriptions(server_id,
                                                      filter_path)
    ```
    The error occurs on the last line, the add_subscriptions.

**Comment by andy-maier** [comment]
On your issue 4: I apparently screwed it in `_createInstance()`. It executes `CreateInstance` twice. Also, we should not set a path anymore, now that we do a `GetInstance`. I'll fix that.

**Comment by andy-maier** [comment]
PR #618 fixes item 4.

**Comment by KSchopmeyer** [comment]
We both have prs up for this since somehow we both took responsibility for this one.

Somehow we have to merge our two prs but since there is only a little overlap, should not be to hard.

**Comment by andy-maier** [comment]
Karl will verify whether all fixes are in now.

**Comment by andy-maier** [comment]
Fixed in PR #619 which was reviewed and merged.
